### PR TITLE
ribosome::api::call()  => ribosome::run_dna()

### PR DESCRIPTION
**Pull request by ddd-mtl**
and moved it in its own file out of 'api' folder.
Move runtime to its own file as well.

There was too much confusion with call() of the hdk api.

### CODE REVIEW

File: core/src/nucleus/ribosome/run_dna.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+/// Executes an exposed zome function in a wasm binary.
+/// Multithreaded function
+/// panics if wasm binary isn't valid.
+pub fn run_dna(
+    dna_name: &str,
+    context: Arc<Context>,
+    wasm: Vec<u8>,
+    zome_call: &ZomeFnCall,
+    parameters: Option<Vec<u8>>,
+) -> ZomeFnResult {
```
Comment (Connoropolous) [inline core/src/nucleus/ribosome/run_dna.rs]: This function is such a monster lol

Comment (Connoropolous) [reply core/src/nucleus/ribosome/run_dna.rs]: Gradually getting better though

File: core/src/nucleus/ribosome/runtime.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        let encoded_allocation = allocation_of_result.unwrap().encode();
+
+        // Return success in i32 format
+        Ok(Some(RuntimeValue::I32(encoded_allocation as i32)))
+    }
+}
+
+// Correlate the indexes of core API functions with a call to the actual function
+// by implementing the Externals trait from Wasmi.
+impl Externals for Runtime {
```
Comment (Connoropolous) [inline core/src/nucleus/ribosome/runtime.rs]: Nice, then I like how this is out here now

Comment (Connoropolous) [reply core/src/nucleus/ribosome/runtime.rs]: I was wondering whether it actually needed to be defined within the function

File: core/src/nucleus/ribosome/run_dna.rs (hunk 1 above)
Comment (ddd-mtl) [reply core/src/nucleus/ribosome/run_dna.rs]: yep. There is no intent to make it smaller, thats why I'm giving it its own file.
I did shrunk it a bit my removing that Runtime impl.

File: core/src/nucleus/ribosome/runtime.rs (hunk 2 above)
Comment (ddd-mtl) [reply core/src/nucleus/ribosome/runtime.rs]: Not 'needed', it just helped scope things a bit.

File: core/src/nucleus/ribosome/run_dna.rs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    // invoke_index and resolve_func work together to enable callable host functions
+    // within WASM modules, which is how the core API functions
+    // read about the Externals trait for more detail
+
+    // Correlate the names of the core ZomeApiFunction's with their indexes
+    // and declare its function signature (which is always the same)
+    struct RuntimeModuleImportResolver;
+    impl ModuleImportResolver for RuntimeModuleImportResolver {
+        fn resolve_func(
```
Comment (Connoropolous) [inline core/src/nucleus/ribosome/run_dna.rs]: @ddd-mtl couldn't we possibly define `RuntimeModuleImportResolver` outside of this function too, along with its `impl ModuleImportResolver...`

Comment (ddd-mtl) [reply core/src/nucleus/ribosome/run_dna.rs]: probably yes. We just took this code from the sample in wasmi

**Comment by ddd-mtl** [comment]
@Connoropolous I was planning on continuing working in this branch :/
But I guess it would be better if I made a branch with my name in it to make it clear.

**Comment by Connoropolous** [comment]
Oops, sorry @ddd-mtl I didn't think folks did that, and was in a habit of deleting them for some reason

**Pull request merged**